src/ingest/src/main/scala/osmesa/functions/osm/package.scala

```diff
@@ -0,0 +1,297 @@
+      "cliff" -> true,
+      "coastline" -> true,
+      "ridge" -> true,
+      "tree_row" -> true
+    ),
+    "office" -> Map(),
+    "piste:type" -> Map(),
+    "place" -> Map(),
+    "playground" -> Map(
+      "balancebeam" -> true,
+      "slide" -> true,
+      "zipwire" -> true
+    ),
+    "power" -> Map(
+      "line" -> true,
+      "minor_line" -> true
+    ),
+    "public_transport" -> Map(
+      "platform" -> true
+    ),
+    "shop" -> Map(),
+    "tourism" -> Map(),
+    "waterway" -> Map(
+      "canal" -> true,
+      "dam" -> true,
+      "ditch" -> true,
+      "drain" -> true,
+      "river" -> true,
+      "stream" -> true,
+      "weir" -> true
+    )
+  )
+
+  private lazy val logger = Logger.getRootLogger
+
+  private val _isArea = (tags: Map[String, String]) =>
+    tags match {
+      case _ if tags.contains("area") && Set("yes", "no", "true", "1").contains(tags("area").toLowerCase) =>
+        Set("yes", "true", "1").contains(tags("area").toLowerCase)
+      case _ =>
+        // see https://github.com/osmlab/id-area-keys (values are inverted)
+        val matchingKeys = tags.keySet.intersect(AREA_KEYS.keySet)
+        matchingKeys.exists(k => !AREA_KEYS(k).contains(tags(k)))
+    }
+
+  val isArea: UserDefinedFunction = udf(_isArea)
+
+  private val _isMultiPolygon = (tags: Map[String, String]) =>
+    tags.contains("type") && Set("boundary", "multipolygon").contains(tags("type").toLowerCase)
+
+  val isMultiPolygon: UserDefinedFunction = udf(_isMultiPolygon)
+
+  val buildWay: UserDefinedFunction = udf((nodes: Seq[Row], isArea: Boolean) => {
+    val geom = nodes
+      .sortWith(_.getAs[Int]("idx") < _.getAs[Int]("idx"))
+      .map(row =>
+        (Option(row.get(row.fieldIndex("lon"))).map(_.asInstanceOf[Double]).getOrElse(Double.NaN),
+          Option(row.get(row.fieldIndex("lat"))).map(_.asInstanceOf[Double]).getOrElse(Double.NaN)))
+    match {
+      // drop ways with invalid coordinates
+      // check for nulls (or add a coalesce to the select after join); since posexplode_outer is used, there will
+      // always be a row but it may be empty
+      case coords if coords.exists { case (x, y) => x.equals(Double.NaN) || y.equals(Double.NaN) } => None
+      // drop ways that don't contain valid geometries
+      case coords if coords.isEmpty => Some("LINESTRING EMPTY".parseWKT)
+      case coords if coords.length == 1 =>
+        Some(Point(coords.head._1, coords.head._2))
+      case coords =>
+        val line = Line(coords)
+
+        Some(isArea match {
+          case true if line.isClosed && coords.length >= 4 => Polygon(line)
+          case _ => line
+        })
+    }
+
+    geom match {
+      // drop invalid geometries
+      case Some(g) if g.isValid => g.toWKB(4326)
+      case _ => null
+    }
+  })
+
+  // create fully-formed rings from line segments
+  @tailrec
+  private def connectSegments(segments: List[Line], rings: List[Polygon] = List.empty[Polygon]): List[Polygon] = {
```

Comment:
The use of `vertices` here produces a lot lot of GC pressure. Suggestions on how better to merge these segments? (Tracking a `List[Array[Point]]` might work, but extracting vertices in the first place copies all of them: https://github.com/locationtech/geotrellis/blob/1a30cc4a5b67b32a5b81c83bc7907dc7c3f062e6/vector/src/main/scala/geotrellis/vector/Line.scala#L94-L103